FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,53 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		context.perform {
+			let fetchRequest: NSFetchRequest = ManagedCache.fetchRequest()
+			do {
+				let result = try self.context.fetch(fetchRequest)
+				if let managedCache = result.first {
+					completion(.found(feed: managedCache.toModels(), timestamp: managedCache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		let context = context
+		deleteCachedFeed { error in
+			if error == nil {
+				context.perform {
+					do {
+						let _ = ManagedCache(context: context, feed: feed, timestamp: timestamp)
+						try self.context.save()
+						completion(nil)
+					} catch {
+						completion(error)
+					}
+				}
+			} else {
+				completion(error)
+			}
+		}
```

Comment:
Calling `deleteCachedFeed` first will make insertion less performant as it would be separated into two chunks of code to run on the context's internal queue. Thus in this case it is better to delete previously stored cache manually and have only one `context.perform` block.

Also using `self` inside this block captures self, which is not needed. To avoid it `context` that was created on line 48 can be used. 

```suggestion
		let context = context
		context.perform {
			do {
				try context.fetch(ManagedCache.fetchRequest()).first.map(context.delete)
				
				let _ = ManagedCache(context: context, feed: feed, timestamp: timestamp)

				try context.save()
				completion(nil)
			} catch {
				completion(error)
			}
		}
```

Also it helps to avoid callback hell and keeps the code concise

Reply:
Thanks! I'm applying the changes however I have some doubts:

1. At first I implemented it the way you mentioned but then I thought that I was duplicating the `delete` functionality inside `insert` so I decided to start calling `delete`, that created the nested closure but I felt that I was not "repeating" code. Does that make any sense for you?.

2. I also tried to do batch delete, but the tests were unable to inject errors properly, I read some docs and found that it is not easy to handle the result of a batch delete. Is there a way to use it here? @danillahtin

Reply:
Hey @zgluis!

1. Yes, I got that from the code. The thing is that duplication is gone but other problems arose (performance, callbacks, bad readability). For this function you only need to remove stored `ManagedCache` if it was persisted. `deleteCachedFeed` not only deletes `ManagedCache`, but also it saves the context. Actually it is a method of another level of abstraction. Potentially in the future `deleteCachedFeed` can appear to have a lot more details which would be redundant or even harmful to use inside insert. There's a way to get rid of duplication without reusing higher level functions: create a helper that will do only the deletion, that's it. Thus you will be able to reuse it everywhere without all the cons of reusing `deleteCachedFeed`.
2. I don't see batch deletion here. We know that `CoreDataFeedStore` operates only on one object. The tests guarantee that we do it properly so that there're no 2 instances of `ManagedCache` at a time in Core Data. Thus it is more efficient to do the minimal amount of work and not to try remove 1 object with batch deletion.